### PR TITLE
Filter functions

### PR DESCRIPTION
**Pull request by lewtun**
#### What does this implement/fix? Explain your changes.
Implements the following filter functions:

* Entropy 
* Gaussian density (see [here](http://danifold.net/mapper/filters.html) for definition)
* Projection

#### Any other comments?
* The Entropy filter is subtle because we must assume that each row in `X` corresponds to a probability. The current approach simply normalises the row values to sum to one, but this is not strictly correct.
* The GaussianDensity filter relies on scikit-learn's KernelDensity estimator which does not have a `transform()` method. The current approach is to simply inject the desired outputs into a new Transformer, but there are likely better ways to do this!
* No need to review this over the weekend :)

### CODE REVIEW

File: giotto/mapper/filter.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            API requires this parameter.
+
+        Returns
+        -------
+        Xt : ndarray, shape (n_samples, 1)
+        """
+        # TODO: the following is a crude method to ensure each row vector
+        # consists of "probabilities" that sum to one. Consider normalisation
+        # in terms of bin counts?
+        X = check_array(X)
```
Comment (ulupo) [inline giotto/mapper/filter.py]: Should there be some exception handling to exclude negative entries? Or should we take abs of X instead? Why neither?

File: giotto/mapper/filter.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        -------
+        Xt : ndarray, shape (n_samples, 1)
+        """
+        kde = KernelDensity(kernel='gaussian',
+                            metric='euclidean', bandwidth=np.std(X)).fit(X)
+        Xt = np.exp(kde.score_samples(X)).reshape(-1, 1)
+        return Xt
+
+
+class Projection(BaseEstimator, TransformerMixin):
```
Comment (ulupo) [inline giotto/mapper/filter.py]: Let's quickly check that this works with lists of integers as well as individual integers

File: giotto/mapper/filter.py (hunk 1 above)
Comment (lewtun) [reply giotto/mapper/filter.py]: Good catch! Instead of throwing an exception, we could raise a warning that negative entries were detected and that we are now taking the absolute value

File: giotto/mapper/filter.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        # TODO: the following is a crude method to ensure each row vector
+        # consists of "probabilities" that sum to one. Consider normalisation
+        # in terms of bin counts?
+        X = check_array(X)
+        probs = X / X.sum(axis=1, keepdims=True)
+        Xt = (entr(probs).sum(axis=1) / np.log(2)).reshape(-1, 1)
+        return Xt
+
+
+class GaussianDensity(BaseEstimator, TransformerMixin):
```
Comment (ulupo) [inline giotto/mapper/filter.py]: Let's consider extending this to a wrapper of the full KernelDensity transformer, accepting the same list of parameters. We could even make it a subclass of KernelDensity and redefine transform to be super().score_samples(X).

More generally, this situation is the same as the one we saw with IsolationForest. There also we wanted score_samples to play the role of transform. Might this be shared with yet more estimators with a score_samples method? If so, we might imagine making a class decorator which modifies the behaviour of all these estimators in the same way.

File: giotto/mapper/filter.py (hunk 1 above)
Comment (ulupo) [reply giotto/mapper/filter.py]: Sounds good!

File: giotto/mapper/tests/test_filter.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+@given(X=arrays(
+    dtype=np.float,
+    elements=floats(allow_nan=False,
+                    allow_infinity=False,
+                    min_value=1,
+                    max_value=1e3),
+    shape=array_shapes(min_dims=2, max_dims=2)
+))
+def test_entropy_values_per_row(X):
```
Comment (ulupo) [inline giotto/mapper/tests/test_filter.py]: Could add a test for handling of negative values...
